src/librepgp/stream-parse.cpp

```diff
@@ -2084,48 +2154,86 @@ init_encrypted_src(pgp_parse_handler_t *handler, pgp_source_t *src, pgp_source_t
             goto finish;
         }
 
-        pgp_key_request_ctx_t keyctx = {};
-        keyctx.op = PGP_OP_DECRYPT_SYM;
-        keyctx.secret = true;
-        keyctx.search.type = PGP_KEY_SEARCH_KEYID;
-
         for (auto &pubenc : param->pubencs) {
-            keyctx.search.by.keyid = pubenc.key_id;
-            /* Get the key if any */
-            if (!(seckey = pgp_request_key(handler->key_provider, &keyctx))) {
-                errcode = RNP_ERROR_NO_SUITABLE_KEY;
-                continue;
-            }
-            /* Decrypt key */
-            if (seckey->encrypted()) {
-                pgp_password_ctx_t pass_ctx{.op = PGP_OP_DECRYPT, .key = seckey};
-                decrypted_seckey =
-                  pgp_decrypt_seckey(*seckey, *handler->password_provider, pass_ctx);
-                if (!decrypted_seckey) {
-                    errcode = RNP_ERROR_BAD_PASSWORD;
-                    continue;
+            const std::array<unsigned char, 8> zeros{
+              0,
+            };
+            if (pubenc.key_id == zeros) {
+                /* RFC 4880, 5.1:
+                 * An implementation MAY accept or use a Key ID of zero as a "wild card"
+                 * or "speculative" Key ID.  In this case, the receiving implementation
+                 * would try all available private keys, checking for a valid decrypted
+                 * session key.  This format helps reduce traffic analysis of messages. */
+                rnp_identifier_iterator_t it = NULL;
+                rnp_key_handle_t          handle = NULL;
+                const char *              identifier = NULL;
+
+                /* Iterate through the keys */
+                rnp_ffi_t ffi = (rnp_ffi_t) handler->key_provider->userdata;
+                if (rnp_identifier_iterator_create(ffi, &it, "keyid")) {
```

Comment:
It would be faster to use fingerprint iterator here and in `rnp_locate_key()`, as in case of preloated keyring keys are stored in the map using the fingerprint as a key.